src/backends/policy_engine/souffle/datalog_lowering_visitor.cc

```diff
@@ -41,7 +42,9 @@ using DatalogAttributePayload = ir::datalog::AttributePayload;
 
 static DatalogAttributePayload GetPayloadForAttribute(ir::Attribute attr,
                                                       ir::SsaNames &ssa_names) {
-  if (auto int_attr = attr.GetIf<ir::Int64Attribute>()) {
+  if (auto double_attr = attr.GetIf<ir::DoubleAttribute>()) {
+    return DatalogAttribute::Number(double_attr->value());
```

Comment:
This should not be `DatalogAttribute::Number`. `Number` is chosen to be parallel to Datalog's `Number` type which is actually its integer type. `Number`'s constructor accepts an `int64_t`. A new type of `DatalogAttribute` called `Float` should be introduced.

I think there must be a gap in testing; this should cause precision to be lost (ie, all the decimal places) which should be fairly easily detectable.

Reply:
Thanks! I'll add one as desired. I'm interested that the `Number` type doesn't require a conversion but will investigate.

Reply:
Ah looking at `souffle/src/include/souffle/TypeAttribute.h` I think it's actually `souffle::TypeAttribute::Signed` that our `Number` type corresponds to.

I'll update the comment to match unless I hear back :)

Reply:
Are you looking inside the Souffle codebase? I was just thinking about the primitive types exposed by Souffle. Maybe they call it `Signed` internally, but in the Datalog files, [it is called `number`](https://souffle-lang.github.io/types#number-type)

Reply:
Ah, thanks, that surprised me. Fixed them now.

Reply:
Ah, I've found that there datalog_lowering_visitor_test has no number tests  and so I didn't discover it when adding Float as an analogue. Adding tests for both now.

Reply:
This still looks to me like it's using DatalogAttribute::Number, which takes a parameter of type int64_t.

Reply:
+1 Sorry for the slow response time, I've been trying to work out why this is building.

Currently, there are no tests for int64 or double attributes lowering to datalog, but the GetIf gives back an ir::FloatAttribute (ptr) and the `value()` call gets a `double`. Additionally the Number constructor is marked as `explicit` so it shouldn't be allowed to cast the double to an int.

Can you illuminate this for me?

Reply:
I think you might be misunderstanding what `explicit` does for C++ constructors.

If you have a single-argument C++ constructor not marked explicit of the form:

```C++
class Foo {
  Foo(Bar bar) /*...*/
};
```

What this means is that `Foo` can be constructed from `Bar` *even as an implicit cast*. Like many cases where C++ is trying to be helpful, this is generally considered a bad move. Fortunately, `explicit` turns it off.

Now in this case, the problem is that the constructor accepts an `int64_t` but you are passing in a `double`. In languages with strong typing, this is considered Very Bad and causes a compiler error. [C++, however, tries to be helpful](https://en.cppreference.com/w/cpp/language/implicit_conversion):

```
A [prvalue](https://en.cppreference.com/w/cpp/language/value_category#prvalue) of floating-point type can be converted to a prvalue of any integer type. The fractional part is truncated, that is, the fractional part is discarded. If the value cannot fit into the destination type, the behavior is undefined (even when the destination type is unsigned, modulo arithmetic does not apply). If the destination type is bool, this is a boolean conversion (see below)
```

Unfortunately this cannot be turned off without going to extreme lengths (like wrapping all primivite types in classes and defining `operator`s to give you stronger typing, a move that seems like a bridge too far for pretty much everyone).